test/Test/Database/LSMTree/Generators.hs

```diff
@@ -119,3 +137,40 @@ labelTestKOps kops' =
     values = foldMap (bifoldMap pure mempty . snd) kops
 
     isLarge = not . uncurry entryWouldFitInPage
+
+prop_withRunDoesntLeak ::
+     FS.HasFS IO h
+  -> FS.HasBlockIO IO h
+  -> SerialisedRunData
+  -> IO Property
+prop_withRunDoesntLeak hfs hbio rd = do
+    let index = Index.Ordinary
+    withRunAt hfs hbio index (simplePath 0) rd $ \_run -> do
+      return (QC.property True)
+
+prop_withMergingRunDoesntLeak ::
+     FS.HasFS IO h
+  -> FS.HasBlockIO IO h
+  -> SerialisedMergingRunData MR.LevelMergeType
+  -> IO Property
+prop_withMergingRunDoesntLeak hfs hbio mrd = do
+    let index = Index.Ordinary
+    let path = FS.mkFsPath []
+    counter <- newUniqCounter 0
+    withMergingRun hfs hbio resolveVal index path counter mrd $ \_mr -> do
+      return (QC.property True)
+
+prop_withMergingTreeDoesntLeak ::
+     FS.HasFS IO h
+  -> FS.HasBlockIO IO h
+  -> SerialisedMergingTreeData
+  -> IO Property
+prop_withMergingTreeDoesntLeak hfs hbio mrd = do
+    let index = Index.Ordinary
+    let path = FS.mkFsPath []
+    counter <- newUniqCounter 0
+    withMergingTree hfs hbio resolveVal index path counter mrd $ \_tree -> do
+      return (QC.property True)
```

Comment:
Maybe add a note/TODO that this only test that nothing is leaked on the happy path, and that we'd have to use `simErrorHasBlockIO` to test nothing is leaked on the unhappy path